tests: stepper_api: create a scaffold to generalize tests

jilaypandya opened the pull request:
This PR creates a scaffold for executing stepper_api tests for the existing step_dir drivers.

### Point of interest:
```
tests/drivers/stepper/stepper_api/testcase.yaml
```
Pinging @jbehrensnx as this test suite should step by step integrate the use-cases from drv8424 api tests.

#### Testing on Hardware

```
west twister \
    --platform nucleo_g071rb/stm32g071xx \    
    -T tests/drivers/stepper/stepper_api \    
    --prep-artifacts-for-testing
```


```
west twister \
        --platform nucleo_g071rb/stm32g071xx \
        -T tests/drivers/stepper/stepper_api \
        --device-testing \
        --device-serial /dev/tty.deadbeef --test-only \
        --west-flash="--skip-rebuild" --west-runner openocd  \
        -v
```
###### Output 
```
===================================================================
RunID: 08d79a9230f36e3debc6c8186f6d5590
PROJECT EXECUTION SUCCESSFUL
D: Initializing motor_1 gpio_stepper with 4 pin
*** Booting Zephyr OS build v4.1.0-269-g7df3874f120e ***
Running TESTSUITE stepper
===================================================================
START - test_actual_position
 PASS - test_actual_position in 0.001 seconds
===================================================================
START - test_get_micro_step_res
 PASS - test_get_micro_step_res in 0.001 seconds
===================================================================
START - test_set_micro_step_interval_invalid_zero
E: Step interval is invalid.
 PASS - test_set_micro_step_interval_invalid_zero in 0.003 seconds
===================================================================
START - test_set_micro_step_res_incorrect
E: Unsupported micro step resolution 127
 PASS - test_set_micro_step_res_incorrect in 0.004 seconds
===================================================================
START - test_stop
D: Event 4, stepper_print_event_callback called for motor_1, expected for motor_1

D: Motor is not moving
 PASS - test_stop in 0.010 seconds
===================================================================
START - test_target_position_w_fixed_step_interval
D: Setting Motor step interval to 100000000
D: Event 0, stepper_print_event_callback called for motor_1, expected for motor_1

 PASS - test_target_position_w_fixed_step_interval in 1.013 seconds
===================================================================
TESTSUITE stepper succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [stepper]: pass = 6, fail = 0, skip = 0, total = 6 duration = 1.032 seconds
 - PASS - [stepper.test_actual_position] duration = 0.001 seconds
 - PASS - [stepper.test_get_micro_step_res] duration = 0.001 seconds
 - PASS - [stepper.test_set_micro_step_interval_invalid_zero] duration = 0.003 seconds
 - PASS - [stepper.test_set_micro_step_res_incorrect] duration = 0.004 seconds
 - PASS - [stepper.test_stop] duration = 0.010 seconds
 - PASS - [stepper.test_target_position_w_fixed_step_interval] duration = 1.013 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
RunID: 08d79a9230f36e3debc6c8186f6d5590
PROJECT EXECUTION SUCCESSFUL
```



## P.S.
This pr is to agree upon a structure as to how we want to use stepper_api tests for all the existing drivers in the tree.

jilaypandya committed:
tests: stepper_api: create a scaffold to generalize tests

This commit creates a scaffold for executing stepper_api tests
on the existing step_dir drivers.

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya committed:
MAINTAINERS: add tests/drivers/stepper to stepper area

Add missing tests/drivers/stepper to "Drivers: Stepper"

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

dipakgmx commented:
Was wondering why not pass the different steppers through additional overlay's via the `testcase.yaml` ?

jilaypandya commented:
> Was wondering why not pass the different steppers through additional overlay's via the `testcase.yaml` ?

That's right, it's the correct way

jilaypandya committed:
tests: stepper_api: test stop, set_micro_step_res, set_micro_step_interval

unify error codes from all drivers based on stepper specification
add a generic test case for set_micro_step_interval and stop

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya commented on tests/drivers/stepper/drv8424/api/src/main.c:
```diff
@@ -79,15 +79,6 @@ ZTEST_F(drv8424_api, test_micro_step_res_set)
 		      res);
 }
 
-ZTEST_F(drv8424_api, test_micro_step_res_set_incorrect_value_detection)
```
Dropping this test case. Since as per the api spec `stepper_set_micro_step_res ` return `-ENOTSUP`. Have added a test case for the same in stepper_api. Have also corrected the drv8424 driver.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -58,46 +78,86 @@ static void stepper_before(void *f)
 ZTEST_SUITE(stepper, NULL, stepper_setup, stepper_before, NULL, NULL);
 
-ZTEST_F(stepper, test_micro_step_res)
+ZTEST_F(stepper, test_set_micro_step_res)
```
Rename Test to make it clear we are checking that incorrect inputs are detected. Maybe append `_incorrect_value` or something similar to the test?

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -58,46 +78,86 @@ static void stepper_before(void *f)
+{
+	int ret = stepper_set_micro_step_res(fixture->dev, 127);
+
+	zassert_equal(ret, -ENOTSUP, "Micro step resolution not set correctly");
```
Change message to better fit with the test. Maybe something like _Incorrect micro step resolution not detected._

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -58,46 +78,86 @@ static void stepper_before(void *f)
+		stepper_is_moving(fixture->dev, &is_moving);
+		zassert_equal(is_moving, true,
+			      "Stepper should be moving since stop is not implemented");
+	}
```
Handle the other possible error codes. Generally, if `ret` is not `0` or `-ENOSYS`, something is wrong and the test should fail.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -16,10 +18,22 @@ struct k_poll_signal stepper_signal;
 struct k_poll_event stepper_event;
 void *user_data_received;
 
+#define POLL_AND_CHECK_SIGNAL(signal, event, expected_event, timeout)                              \
```
Why not use a normal helper function? This would alleviate the need for the `do while(0)` construct.

jilaypandya replied:
If you take a normal helper function that in your test log the failing line would be that of the function. 

So in the case that you use this function multiple times in a test case you will be not really able to pinpoint where the error took place.

jbehrensnx replied:
In that case enclose the curly brackets with regular brackets i.e. `({ code })` instead. This will have the same effect and is easier to read.

jbehrensnx commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -11,3 +11,30 @@ tests:
       - nucleo_g071rb
       - nucleo_f767zi
       - qemu_x86_64
+  drivers.stepper.stepper_api.adi_tmc2209:
```
Possibly add native_sim in some form to the name of these test suites. While it is the only platform allowed, it would also make this clear by their names that these are native_sim tests.

The current name implies that they are normal test suites for these drivers .

Maybe `.adi_tmc2209.native_sim` or `.native_sim.adi_tmc2209` , which would clarify this and also open an obvious entry point for doing the same for other boards.

I'm very much open to discussion on this point.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -58,46 +78,88 @@ static void stepper_before(void *f)
+		      "Micro step resolution not set correctly");
+}
+
+ZTEST_F(stepper, test_set_micro_step_interval)
```
Adjust name to clarify, that we check that ms-interval 0 throws an error.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -58,46 +78,88 @@ static void stepper_before(void *f)
+	int32_t pos = 10;
 
-	(void)stepper_set_microstep_interval(fixture->dev, 10000);
+	(void)stepper_set_microstep_interval(fixture->dev, NSEC_PER_SEC);
```
10s for this test case is way to long. Maybe consider increasing speed to 10 steps/s. In my experience, native_sim can easily do this.

That said, it might be worthwhile to disable this test for the time being. The reason for this are ramp drivers. One can not be sure that the simple timeout calculation is sufficient for them. This is the reason why I moved this test to the const-velocity test suite in my test-generalization PR.

jilaypandya replied:
> That said, it might be worthwhile to disable this test for the time being. The reason for this are ramp drivers. One can not be sure that the simple timeout calculation is sufficient for them. This is the reason why I moved this test to the const-velocity test suite in my test-generalization PR.

This test has been part of api suite and let's remove or modify this once we have something for ramp drivers :)

jbehrensnx replied:
Genuine curiosity: does the tmc50xx driver successfully complete the test suite? That is the main reason why I made that comment.

jilaypandya replied:
I have renamed the test to `test_target_position_w_fixed_step_interval `, this test will have to be skipped for ramp drivers like tmc50xx

jilaypandya commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -11,3 +11,30 @@ tests:
+      - drivers
+      - stepper
+      - api
+    extra_args: DTC_OVERLAY_FILE="boards/native_sim_adi_tmc2209.overlay" CONFIG_GPIO=y
```
Thanks :)

faxe1008 commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -11,3 +11,40 @@ tests:
       - qemu_x86_64
+  drivers.stepper.stepper_api.adi_tmc2209:
+    tags:
+      - drivers
```
Not too sure but could you move the tags up top to common?

faxe1008 commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -11,3 +11,40 @@ tests:
+      - stepper
+      - api
+    extra_args:
+      - platform:native_sim/native/64:DTC_OVERLAY_FILE="boards/native_sim_adi_tmc2209.overlay"
```
Is the plattform in here needed, if you are filtering the test with plattform_allow anyways?

jilaypandya replied:
Yep. This list might get extended with `platform:nucleo_g071rb:DTC_OVERLAY_FILE="boards/nucleo_g071rb_adi_tmc2209.overlay` in the future :)

faxe1008 replied:
Not to sure :^), not an twister expert by any means. Have not seen that construct anywhere else tho. I would add another entry to the sample for specific configurations that are not covered by a "generalized" board overlay. But maybe some of the twister folks can comment on that one :^) (@golowanow @nashif )

jilaypandya replied:
https://github.com/zephyrproject-rtos/zephyr/blob/c4ebf713b10ce63f44ee3497159c54377ecc839b/samples/drivers/display/sample.yaml#L100

jbehrensnx replied:
It might be a good idea to have qemu switch to this model as well. Like native_sim it is an emulation/simulation platform so multiple drivers can easily be tested, and it would make the advantages of this scaffold system immediately clear and why the platform tag is used for the overlay.

jilaypandya replied:
done
